File: mycollab-caching/src/main/java/com/mycollab/cache/service/impl/InfinispanCacheService.java
/**
 * This file is part of mycollab-caching.
 *
 * mycollab-caching is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-caching is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-caching.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.cache.service.impl;

import com.mycollab.cache.service.CacheService;
import org.apache.commons.collections.CollectionUtils;
import org.infinispan.AdvancedCache;
import org.infinispan.commons.api.BasicCache;
import org.infinispan.context.Flag;
import org.infinispan.manager.DefaultCacheManager;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Service;

import java.io.InputStream;
import java.util.Set;

/**
 * @author MyCollab Ltd
 * @since 5.1.4
 */
@Service
@Order(value = 1)
public class InfinispanCacheService implements CacheService, InitializingBean {
    private static Logger LOG = LoggerFactory.getLogger(InfinispanCacheService.class);

    private DefaultCacheManager instance;

    @Override
    public void afterPropertiesSet() throws Exception {
        try {
            InputStream configInputStream;
            configInputStream = InfinispanCacheService.class.getClassLoader().getResourceAsStream("infinispan-local.xml");

            if (configInputStream == null) {
                configInputStream = InfinispanCacheService.class.getClassLoader().getResourceAsStream("infinispan-default.xml");
            }
            instance = new DefaultCacheManager(configInputStream);

        } catch (Exception e) {
            LOG.debug("Error while set up infinispan cache manager. Will initiate the default", e);
            instance = new DefaultCacheManager();
        }
    }

    @Override
    public void putValue(String group, String key, Object value) {
        BasicCache<String, Object> cache = getCache(group);
        cache.put(key, value);
    }

    private BasicCache<String, Object> getCache(String group) {
        BasicCache<String, Object> cache = instance.getCache(group);
        if (cache instanceof AdvancedCache) {
            cache = ((AdvancedCache<String, Object>) cache).withFlags(Flag.IGNORE_RETURN_VALUES);
        }
        return cache;
    }

    @Override
    public Object getValue(String group, String key) {
        BasicCache<String, Object> cache = getCache(group);
        return cache.get(key);
    }

    @Override
    public void removeCacheItem(String id, String prefixKey) {
        BasicCache<String, Object> cache = instance.getCache(id);
        LOG.debug("Remove cache has prefix {} in group {}", prefixKey, id);
        Set<String> keys = cache.keySet();
        if (CollectionUtils.isNotEmpty(keys)) {

            String[] keyArr = keys.toArray(new String[0]);
            for (int i = 0; i < keyArr.length; i++) {
                if (keyArr[i].startsWith(prefixKey)) {
                    LOG.debug("Remove cache key {}", keyArr[i]);
                    cache.remove(keyArr[i]);
                }
            }
        }
    }

    @Override
    public void removeCacheItems(String group, Class<?>... classes) {
        for (Class<?> prefKey : classes) {
            removeCacheItem(group, prefKey.getName());
        }
    }
}

File: mycollab-caching/src/main/java/com/mycollab/cache/service/impl/InfinispanCacheService.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.cache.service.impl

import com.mycollab.cache.service.CacheService
import org.infinispan.AdvancedCache
import org.infinispan.commons.api.BasicCache
import org.infinispan.context.Flag
import org.infinispan.manager.EmbeddedCacheManager
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

/**
 * @author MyCollab Ltd
 * @since 5.1.4
 */
@Service
class InfinispanCacheService(private val instance: EmbeddedCacheManager) : CacheService {

    override fun putValue(group: String, key: String, value: Any) {
        val cache = getCache(group)
        cache[key] = value
    }

    private fun getCache(group: String): BasicCache<String, Any> {
        var cache: BasicCache<String, Any> = instance.getCache(group)
        if (cache is AdvancedCache<*, *>) {
            cache = (cache as AdvancedCache<String, Any>).withFlags(Flag.IGNORE_RETURN_VALUES)
        }
        return cache
    }

    override fun getValue(group: String, key: String): Any? {
        val cache = getCache(group)
        return cache[key]
    }

    override fun removeCacheItem(group: String, prefixKey: String) {
        val cache = instance.getCache<String, Any>(group)
        LOG.debug("Remove cache has prefix $prefixKey in group $group")
        val keys = cache.keys
        keys.forEach {
            if (it.startsWith(prefixKey)) {
                LOG.debug("Remove cache key $it")
                cache.remove(it)
            }
        }
    }

    override fun removeCacheItems(group: String, vararg classes: Class<*>) {
        classes.forEach { removeCacheItem(group, it.name) }
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(InfinispanCacheService::class.java)
    }
}
